File: src/main/java/fun/vs/fw/demo/jpa/CategorizedTransactionEntity.java
package fun.vs.fw.demo.jpa;

import fun.vs.fw.demo.domain.CategorizedTransaction;
import fun.vs.fw.demo.domain.CategorizedTransaction.CategorizedTransactionId;
import fun.vs.fw.demo.domain.CategorizedTransaction.ExpenseCategory;
import fun.vs.fw.demo.domain.Transaction.AccountId;
import fun.vs.fw.demo.domain.Transaction.ClientId;
import fun.vs.fw.demo.domain.Transaction.Amount;
import fun.vs.fw.demo.domain.Transaction.TransactionId;
import jakarta.persistence.*;

import java.math.BigDecimal;

/**
 * Represents a categorized financial transaction associated with a specific client and account.
 * This entity includes attributes such as the transaction ID, client ID, account ID, transaction
 * amount, and the expense category to which the transaction belongs.
 * <p>
 * This class is marked as a JPA entity and mapped to the "categorized_transaction" table
 * in the database.
 */
@Table(name = "categorized_transaction")
@Entity public class CategorizedTransactionEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "transaction_id", unique = true, nullable = false)
    private String transactionId;

    @Column(name = "client_id", nullable = false)
    private String clientId;

    @Column(name = "account_id", nullable = false)
    private String accountId;

    @Column(name = "amount", nullable = false, precision = 2)
    private BigDecimal amount;

    @Column(name = "expense_category", nullable = false)
    private String expenseCategory;

    public CategorizedTransactionEntity() {
    }

    public CategorizedTransactionEntity(Long id, String transactionId, String clientId, String accountId, BigDecimal amount, String expenseCategory) {
        this(transactionId, clientId, accountId, amount, expenseCategory);
        this.id = id;
    }

    public CategorizedTransactionEntity(String transactionId, String clientId, String accountId, BigDecimal amount, String expenseCategory) {
        this.transactionId = transactionId;
        this.clientId = clientId;
        this.accountId = accountId;
        this.amount = amount;
        this.expenseCategory = expenseCategory;
    }

    public static CategorizedTransactionEntity valueOf(CategorizedTransaction categorizedTransaction) {
        return new CategorizedTransactionEntity(
                categorizedTransaction.id().value(),
                categorizedTransaction.transactionId().value(),
                categorizedTransaction.clientId().value(),
                categorizedTransaction.accountId().value(),
                categorizedTransaction.amount().value(),
                categorizedTransaction.expenseCategory().value()
        );
    }

    public CategorizedTransaction toDomain() {
        return new CategorizedTransaction(
                new CategorizedTransactionId(id),
                new TransactionId(transactionId),
                new ClientId(clientId),
                new AccountId(accountId),
                new Amount(amount),
                new ExpenseCategory(expenseCategory)
        );
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getTransactionId() {
        return transactionId;
    }

    public void setTransactionId(String transactionId) {
        this.transactionId = transactionId;
    }

    public String getClientId() {
        return clientId;
    }

    public void setClientId(String clientId) {
        this.clientId = clientId;
    }

    public String getAccountId() {
        return accountId;
    }

    public void setAccountId(String accountId) {
        this.accountId = accountId;
    }

    public BigDecimal getAmount() {
        return amount;
    }

    public void setAmount(BigDecimal amount) {
        this.amount = amount;
    }

    public String getExpenseCategory() {
        return expenseCategory;
    }

    public void setExpenseCategory(String expenseCategory) {
        this.expenseCategory = expenseCategory;
    }
}

File: fvf4k/src/main/kotlin/fvf4k/demo/infra/jpa/CategorizedTransactionEntity.kt
package fvf4k.demo.infra.jpa

import arrow.core.NonEmptyList
import arrow.core.raise.ExperimentalRaiseAccumulateApi
import arrow.core.raise.RaiseDSL
import arrow.core.raise.context.Raise
import arrow.core.raise.context.RaiseAccumulate
import arrow.core.raise.context.accumulate
import arrow.core.raise.context.accumulating
import arrow.core.raise.context.withError
import fvf4k.demo.domain.failure.CategorizedTransactionCorrupted
import fvf4k.demo.domain.model.AccountId
import fvf4k.demo.domain.model.CategorizedTransaction
import fvf4k.demo.domain.model.CategorizedTransactionId
import fvf4k.demo.domain.model.ClientId
import fvf4k.demo.domain.model.ExpenseCategory
import fvf4k.demo.domain.model.MerchantCategoryCode
import fvf4k.demo.domain.model.Money
import fvf4k.demo.domain.model.Transaction
import fvf4k.demo.domain.model.TransactionId
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.math.BigDecimal
import java.util.*
import kotlin.experimental.ExperimentalTypeInference
import arrow.core.raise.zipOrAccumulate as zipOrAccumulateExt


@Table(name = "categorized_transaction")
@Entity
data class CategorizedTransactionEntity(
    @Id @GeneratedValue(strategy = GenerationType.UUID) val id: UUID,
    @Column(name = "transaction_id", nullable = false) val transactionId: UUID?,
    @Column(name = "client_id", nullable = false) val clientId: UUID?,
    @Column(name = "account_id", nullable = false) val accountId: UUID?,
    @Column(name = "amount", nullable = false) val amount: BigDecimal?,
    @Column(name = "currencyCode", nullable = false) val currencyCode: String?,
    @Column(name = "mcc", nullable = false) val mcc: String?,
    @Column(name = "expense_category", nullable = false) val expenseCategory: String?
)

@OptIn(ExperimentalRaiseAccumulateApi::class)
context(_: Raise<CategorizedTransactionCorrupted>)
fun CategorizedTransactionEntity.toDomain(): CategorizedTransaction =
    withError({ CategorizedTransactionCorrupted(it) }) {
        accumulate {
            val validTransactionId by accumulating { TransactionId(transactionId) }
            val validClientId by accumulating { ClientId(clientId) }
            val validAccountId by accumulating { AccountId(accountId) }
            val validMoney by accumulating { Money(amount, currencyCode) }
            val validMcc by accumulating { MerchantCategoryCode(mcc) }
            val validExpenseCategory by accumulating { ExpenseCategory(expenseCategory) }

            CategorizedTransaction(
                id = CategorizedTransactionId(id),
                transaction = Transaction(
                    id = validTransactionId,
                    clientId = validClientId,
                    accountId = validAccountId,
                    money = validMoney,
                    mcc = validMcc
                ),
                expenseCategory = validExpenseCategory
            )
        }
    }

fun CategorizedTransaction.toJpaEntity(): CategorizedTransactionEntity =
    CategorizedTransactionEntity(
        id = this.id.value,
        transactionId = this.transaction.id.value,
        clientId = this.transaction.clientId.value,
        accountId = this.transaction.accountId.value,
        amount = this.transaction.money.value,
        currencyCode = this.transaction.money.currency.currencyCode,
        mcc = this.transaction.mcc.value,
        expenseCategory = this.expenseCategory.value
    )
